File: chapter-08-service-graph/ts/src/main.mjs
import path from "node:path";
import process from "node:process";
import {
  formatGraphDiff,
  formatTextGraph,
  listScenarios,
  loadScenario,
  validateScenarios,
} from "./lib.mjs";

function printUsage() {
  process.stdout.write(
    "Usage:\n" +
      "  node ts/src/main.mjs list\n" +
      "  node ts/src/main.mjs render <lab-name> [text|json]\n" +
      "  node ts/src/main.mjs validate [lab-name ...]\n" +
      "  node ts/src/main.mjs diff <from-lab> <to-lab>\n",
  );
}

async function main() {
  const rootDir = path.resolve(path.dirname(new URL(import.meta.url).pathname), "..", "..");
  const command = process.argv[2] ?? "render";

  if (command === "--help" || command === "-h") {
    printUsage();
    return;
  }

  if (command === "list") {
    for (const scenarioName of await listScenarios(rootDir)) {
      process.stdout.write(`${scenarioName}\n`);
    }
    return;
  }

  if (command === "render") {
    const scenarioName = process.argv[3] ?? "lab1-upload-only";
    const format = process.argv[4] ?? "text";
    const { graph } = await loadScenario(rootDir, scenarioName);

    if (format === "json") {
      process.stdout.write(`${JSON.stringify({ scenario: scenarioName, ...graph }, null, 2)}\n`);
      return;
    }

    process.stdout.write(formatTextGraph(graph, scenarioName));
    return;
  }

  if (command === "validate") {
    const scenarioNames = process.argv.slice(3);
    for (const line of await validateScenarios(rootDir, scenarioNames)) {
      process.stdout.write(`${line}\n`);
    }
    return;
  }

  if (command === "diff") {
    const fromScenario = process.argv[3];
    const toScenario = process.argv[4];
    if (!fromScenario || !toScenario) {
      throw new Error("Usage: node ts/src/main.mjs diff <from-lab> <to-lab>");
    }
    process.stdout.write(await formatGraphDiff(rootDir, fromScenario, toScenario));
    return;
  }

  throw new Error(`Unknown command "${command}"`);
}

main().catch((error) => {
  process.stderr.write(`${error.message}\n`);
  process.exitCode = 1;
});

File: chapter-08-service-graph/src/main.rs
use chapter8_service_graph::{
    diff_graphs, format_graph, format_graph_diff, list_scenarios, load_scenario_graph, project_root,
};

fn print_usage() {
    println!("Usage:");
    println!("  cargo run --quiet -- list");
    println!("  cargo run --quiet -- render <scenario> [text|json]");
    println!("  cargo run --quiet -- validate [scenario ...]");
    println!("  cargo run --quiet -- graph-diff <from-scenario> <to-scenario>");
}

fn main() {
    if let Err(error) = run() {
        eprintln!("{error}");
        std::process::exit(1);
    }
}

fn run() -> Result<(), String> {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let root = project_root();

    match args.first().map(String::as_str) {
        Some("--help") | Some("-h") | None => {
            print_usage();
            Ok(())
        }
        Some("list") => {
            for scenario in list_scenarios(&root)? {
                println!("{scenario}");
            }
            Ok(())
        }
        Some("render") => {
            let scenario = args.get(1).map(String::as_str).unwrap_or("lab1-upload-only");
            let format = args.get(2).map(String::as_str).unwrap_or("text");
            let report = load_scenario_graph(&root, scenario)?;
            if format == "json" {
                println!(
                    "{}",
                    serde_json::to_string_pretty(&report).map_err(|err| err.to_string())?
                );
            } else {
                print!("{}", format_graph(&report));
            }
            Ok(())
        }
        Some("validate") => {
            let scenarios = if args.len() > 1 {
                args[1..].to_vec()
            } else {
                list_scenarios(&root)?
            };
            for scenario in scenarios {
                let report = load_scenario_graph(&root, &scenario)?;
                println!(
                    "Validated {}: {} services, {} edges, {} waiting consumers",
                    report.scenario,
                    report.graph.services.len(),
                    report.graph.edges.len(),
                    report.graph.waiting.len()
                );
            }
            Ok(())
        }
        Some("graph-diff") => {
            let from = args
                .get(1)
                .ok_or_else(|| "Usage: cargo run --quiet -- graph-diff <from-scenario> <to-scenario>".to_string())?;
            let to = args
                .get(2)
                .ok_or_else(|| "Usage: cargo run --quiet -- graph-diff <from-scenario> <to-scenario>".to_string())?;
            let left = load_scenario_graph(&root, from)?;
            let right = load_scenario_graph(&root, to)?;
            print!("{}", format_graph_diff(&diff_graphs(&left, &right)));
            Ok(())
        }
        Some(other) => Err(format!("unknown command \"{other}\"")),
    }
}
